File: SlidingWindow/CheckInclusion.java
package SlidingWindow;

public class CheckInclusion {
  public boolean checkInclusion(String s1, String s2) {
    int[] reqCharCounts = new int[26];
    int[] windowCharCounts = new int[26];
    int reqCharsRemaining = 0;

    for (int i = 0; i < s1.length(); i++) {
      int index = getCharIndex(s1.charAt(i));
      if (reqCharCounts[index] == 0)
        reqCharsRemaining++;
      reqCharCounts[index]++;
    }

    int left = 0;
    for (int right = 0; right < s2.length(); right++) {
      int charIndex = getCharIndex(s2.charAt(right));
      windowCharCounts[charIndex]++;
      if (windowCharCounts[charIndex] == reqCharCounts[charIndex])
        reqCharsRemaining--;

      if (reqCharsRemaining == 0)
        return true;

      // Maintain window size
      if (right - left + 1 > s1.length()) {
        charIndex = getCharIndex(s2.charAt(left));
        windowCharCounts[charIndex]--;
        if (windowCharCounts[charIndex] == reqCharCounts[charIndex] - 1)
          reqCharsRemaining++;
      }
    }
    
    return false;
  }

  private int getCharIndex(char letter) {
    return letter - 'a';
  }
}

File: SlidingWindow/checkInclusion.js
const checkInclusion = (s1, s2) => {
  const reqCharCounts = new Map();
  const reqCharsInWindow = new Map();
  let distinctCharsRemaining = 0;

  for (let char of s1) {
    reqCharCounts.set(char, (reqCharCounts.get(char) || 0) + 1);
    if (reqCharCounts.get(char) === 1) distinctCharsRemaining++;
  }

  let left = 0;
  for (let right = 0; right < s2.length; right++) {
    const rightChar = s2[right];
    if (reqCharCounts.get(rightChar)) {
      reqCharsInWindow.set(rightChar, (reqCharsInWindow.get(rightChar) || 0) + 1);
      if (reqCharCounts.get(rightChar) === reqCharsInWindow.get(rightChar))
        distinctCharsRemaining--;
    }

    // Maintain valid window size
    if (right - left + 1 > s1.length) {
      const leftChar = s2[left];
      if (reqCharsInWindow.has(leftChar)) {
        reqCharsInWindow.set(leftChar, reqCharsInWindow.get(leftChar) - 1);
        if (reqCharsInWindow.get(leftChar) === reqCharCounts.get(leftChar) - 1)
          distinctCharsRemaining++;
      }
      left++;
    }

    if (distinctCharsRemaining === 0) return true;
  }

  return false;
};
